src/shims/windows/handle.rs

```diff
@@ -0,0 +1,118 @@
+use rustc_target::abi::HasDataLayout;
+use std::mem::variant_count;
+
+use crate::*;
+
+/// A Windows `HANDLE` that represents a resource instead of being null or a pseudohandle.
+///
+/// This is a seperate type from [`Handle`] to simplify the packing and unpacking code.
+#[derive(Clone, Copy)]
+enum RealHandle {
+    Thread(ThreadId),
+}
+
+impl RealHandle {
+    fn discriminant(self) -> u64 {
+        match self {
+            // can't use zero here because all zero handle is invalid
+            Self::Thread(_) => 1,
+        }
+    }
+
+    fn data(self) -> u64 {
+        match self {
+            Self::Thread(thread) => thread.to_u32() as u64,
+        }
+    }
+
+    fn packed_disc_size() -> u64 {
+        (variant_count::<Self>().log2() + 1) as u64
+    }
+
+    fn to_packed(self, bits: u64) -> u64 {
+        // top bit and lower 2 bits need to be clear
+        let usable_bits = (bits - 3).min(32);
+
+        let disc_size = Self::packed_disc_size();
+        let data_size = usable_bits - disc_size;
+
+        let discriminant = self.discriminant();
+        let data = self.data();
+
+        assert!(discriminant < 2u64.pow(disc_size as u32));
+        assert!(data < 2u64.pow(data_size as u32));
+
+        (discriminant << data_size | data) << 2
+    }
+
+    fn new(discriminant: u32, data: u32) -> Option<Self> {
+        match discriminant {
+            1 => Some(Self::Thread(data.into())),
+            _ => None,
+        }
+    }
+
+    fn from_packed(handle: u64, bits: u64) -> Option<Self> {
+        let usable_bits = (bits - 3).min(32);
+
+        let disc_size = Self::packed_disc_size();
+        let data_size = usable_bits - disc_size;
+        let data_mask = 2u64.pow(data_size as u32) - 1;
+
+        let discriminant = handle >> data_size >> 2;
+        let data = handle >> 2 & data_mask;
+
+        Self::new(discriminant.try_into().unwrap(), data.try_into().unwrap())
+    }
+}
+
+/// Miri representation of a Windows `HANDLE`
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Handle {
+    Null, // = 0
+    // pseudo-handles
+    CurrentThread, // = -2
+    // real handles
+    Thread(ThreadId),
+}
+
+impl Handle {
+    fn to_packed(self, bits: u64) -> i64 {
+        match self {
+            Self::Null => 0,
+            Self::CurrentThread => -2,
```

Comment:
Please no magic constants. This -2 should have a name that is used in the two relevant places.